core/controllers/base.py

```diff
@@ -201,6 +201,14 @@ def __init__(self, request, response):  # pylint: disable=super-init-not-called
                     user_settings.profile_picture_data_url)
                 if user_settings.last_started_state_editor_tutorial:
                     self.has_seen_editor_tutorial = True
+                # In order to avoid too many datastore writes, we do not bother
+                # recording a log-in if the current time is sufficiently close
+                # to the last log-in time.
+                if (user_settings.last_logged_in is None or
+                        utils.are_datetimes_close(
+                            datetime.datetime.utcnow(),
+                            user_settings.last_logged_in)):
+                    user_services.record_user_logged_in(self.user_id)
```

Comment:
utils.are_datetimes_close() function will be true when two times are separated by time less than `feconf.PROXIMAL_TIMEDELTA_SECS`. In that case we do not want to record the last logged in time. We do it only if they are separated by 12 hours, right ? (I feel like that should be correct logic).

But here we are updating last log in time when, either it has not been recorded previously (we should record) or difference between current time and last log in time of user is less than 12 hour (we should leave it unaffected). So condition is incorrect.
Or may be you are trying some different logic ?